.evergreen/.evg.yml

```diff
@@ -288,6 +288,176 @@ functions:
 
+  "add aws auth variables to file":
+    - command: shell.exec
+      type: test
+      params:
+        working_dir: "src"
+        silent: true
+        script: |
+          cat <<EOF > ${DRIVERS_TOOLS}/.evergreen/auth_aws/aws_e2e_setup.json
+          {
+              "iam_auth_ecs_account" : "${iam_auth_ecs_account}",
+              "iam_auth_ecs_secret_access_key" : "${iam_auth_ecs_secret_access_key}",
+              "iam_auth_ecs_account_arn": "arn:aws:iam::557821124784:user/authtest_fargate_user",
+              "iam_auth_ecs_cluster": "${iam_auth_ecs_cluster}",
+              "iam_auth_ecs_task_definition": "${iam_auth_ecs_task_definition}",
+              "iam_auth_ecs_subnet_a": "${iam_auth_ecs_subnet_a}",
+              "iam_auth_ecs_subnet_b": "${iam_auth_ecs_subnet_b}",
+              "iam_auth_ecs_security_group": "${iam_auth_ecs_security_group}",
+
+              "iam_auth_assume_aws_account" : "${iam_auth_assume_aws_account}",
+              "iam_auth_assume_aws_secret_access_key" : "${iam_auth_assume_aws_secret_access_key}",
+              "iam_auth_assume_role_name" : "${iam_auth_assume_role_name}",
+
+              "iam_auth_ec2_instance_account" : "${iam_auth_ec2_instance_account}",
+              "iam_auth_ec2_instance_secret_access_key" : "${iam_auth_ec2_instance_secret_access_key}",
+              "iam_auth_ec2_instance_profile" : "${iam_auth_ec2_instance_profile}"
+          }
+          EOF
+
+  "configure regular aws credentials":
+    - command: shell.exec
+      type: test
+      params:
+        working_dir: "src"
+        script: |
+          ${PREPARE_SHELL}
+          cd ${DRIVERS_TOOLS}/.evergreen/auth_aws
+          ${MONGODB_BINARIES}/mongo --verbose aws_e2e_regular_aws.js
+          cd -
+
+  "run aws auth test with regular aws credentials":
+    - command: shell.exec
+      type: test
+      params:
+        working_dir: "src"
+        silent: true
+        script: |
+          # DO NOT ECHO WITH XTRACE (which PREPARE_SHELL does)
+
+          alias urlencode='python -c "import sys, urllib as ul; print ul.quote_plus(sys.argv[1])"'
+          USER=$(urlencode ${iam_auth_ecs_account})
+          PASS=$(urlencode ${iam_auth_ecs_secret_access_key})
+
+          JDK=${JDK} MONGODB_URI="mongodb://$USER:$PASS@localhost" .evergreen/run-mongodb-aws-test.sh
+
+  "configure assume role credentials":
+    - command: shell.exec
+      type: test
+      params:
+        working_dir: "src"
+        silent: true
+        script: |
+          # DO NOT ECHO WITH XTRACE (which PREPARE_SHELL does)
+
+          export DRIVERS_TOOLS="$DRIVERS_TOOLS"
+          export MONGODB_BINARIES="$MONGODB_BINARIES"
+
+          cd ${DRIVERS_TOOLS}/.evergreen/auth_aws
+          ${MONGODB_BINARIES}/mongo --verbose aws_e2e_assume_role.js
+          cd -
+
+  "run aws auth test with assume role credentials":
+    - command: shell.exec
+      type: test
+      params:
+        working_dir: "src"
+        silent: true
```

Comment:
I understand using the `silent` option to remove the risk of leaking secrets but it also makes it impossible to know what happened when the tests fail. An alternative is what we did for testing FLE in pymongo. We silently create a file with secrets like this:
```
    - command: shell.exec
      type: test
      params:
        silent: true
        working_dir: "src"
        script: |
          if [ -n "${test_encryption}" ]; then
            cat <<EOT > fle_aws_creds.sh
            export FLE_AWS_KEY="${fle_aws_key}"
            export FLE_AWS_SECRET="${fle_aws_secret}"
          EOT
          fi
```
https://github.com/mongodb/mongo-python-driver/blob/3.10.1/.evergreen/config.yml#L352-L363


Then we can run the tests with `silent: false` and xtrace off by sourcing (and removing) the file in a subsequent `shell.exec`:
```
    - command: shell.exec
      type: test
      params:
...
          if [ -n "${test_encryption}" ]; then
            # Disable xtrace (just in case it was accidentally set).
            set +x
            . ./fle_aws_creds.sh
            rm -f ./fle_aws_creds.sh
            export LIBMONGOCRYPT_URL="${libmongocrypt_url}"
            export TEST_ENCRYPTION=1
          fi
```
https://github.com/mongodb/mongo-python-driver/blob/3.10.1/.evergreen/config.yml#L380-L387 

This way we just need to ensure that the driver (and run-mongodb-aws-*.sh scripts) will never leak the password to stdout/stderr.

I think we could do something similar here but up to you.

Reply:
I'm actually not seeing a difference. I wonder if it's because the tasks that are silent only run the run-mongodb-aws-*.sh scripts and don't do much else. Will try to double-check once SERVER-46198 is resolved since tests are failing now.

Reply:
@bazile-clyde have you had a chance to look again?

Matt says: "It's because Shane's example doesn't do set -x after he does a set +x"

Reply:
Let's decide one way or the other what to do about this.  I expect other driver engineers to copy this configuration, so we should ensure that it's as usable as possible.

Reply:
Hey, I think I've got this refactored in a nice way for node, and we should consider using it here. 

Here are the changes for the config.yml:
https://github.com/mongodb/node-mongodb-native/blob/NODE-2288/mongodb-aws-auth-support/.evergreen/config.yml#L159-L304

And here are the changes for `run-mongodb-aws-test.sh`:
https://github.com/mongodb/node-mongodb-native/blob/NODE-2288/mongodb-aws-auth-support/.evergreen/run-mongodb-aws-test.sh#L17-L19

In particular this does a few things:
- combines the "setup" tasks with the "run tests" tasks they are setting up for
- breaks each task into three parts: setup (not silent), secret/auth provisioning (silent), running tests (not silent)

The benefit of making the first and third steps not silent is that you can see issues with your language-specific test runner, as well as any issues we might have running Mark's mongo shell scripts for setup. This helped me identify that the "assume role" tests were flaky, because they occasionally fail the initial mongo setup with a MongoDB server error.

Reply:
Updated.

Reply:
This still has `silent: true` so we don't get to see any output from run-mongodb-aws-test.sh. I made a suggestion to change it below (and another one on "run aws auth test with regular aws credentials" above).

Reply:
For assume role request and regular credentials the URI contains the secret. The java command to run the test in run-mongodb-aws-test.sh outputs whatever URI it's given. This isn't an issue for test cases since no creds are in the URI but it _is_ an issue for the two cases mentioned above.